### ISSUE TITLE
Leapfrog View Export

### ISSUE THREAD

**Issue by banesullivan** [pyvista/pyvista#163]
We should find a way to export `vtki` scenes to [Steno3D](https://steno3d.com) so that users can rapidly share interactive visualizations of their data over the web.

- https://github.com/seequent/steno3dpy

I think this would be really similar to the VTKjs export script I added in #57 

Pinging @fwkoch: do you think we could collaborate on this? Would this be reasonable to "export" a scene to the Steno3D format and provide an API in `vtki` to send scenes directly to a users account?

**Comment by banesullivan** [comment]
Also - [`omfvtk`](https://github.com/OpenGeoVis/omfvtk) is `vtki` at its core... this functionality could be of interest to create a way to visualize OMF datasets locally using `vtki`, then seamlessly share on the cloud by exporting to Steno3D

**Comment by fwkoch** [comment]
Hi @banesullivan - thanks for mentioning me! This sounds really cool and collaborating would be great.

Rather than integrating with Steno3D, I'd recommend moving over to [LF View](https://lfview.com). This is a rebranded evolution of the same product. It is built on better technology and under active development, whereas Steno3D is no longer maintained (and likely Steno3D users will be transitioned to LF View sometime relatively soon). LF View has all the same features as Steno3D and more. It also has a cleaner REST API with a faster, better, though less mature [Python client library](https://github.com/seequent/lfview-api-client).

Beyond the Steno3D/LF View discussion, it would be totally reasonable to export a scene from `vtki`. Authentication is easy - it just uses an API key. The `omf`/`steno3d`/`lfview` APIs all have fairly similar data structures, so I don't see any major hurdles there. Especially since you've already worked through `omf` -> `vtki` conversion, this should mostly just be translating in the opposite direction.

Let me know if you have follow-up questions/concerns. Happy to contribute however I can.

**Comment by banesullivan** [comment]
Thanks for your swift feedback, @fwkoch!

> I'd recommend moving over to LF View. This is a rebranded evolution of the same product. It is built on better technology and under active development

This looks like an awesome project - I'll work towards building an export script for LF View instead!

> it would be totally reasonable to export a scene from vtki ... The omf/steno3d/lfview APIs all have fairly similar data structures, so I don't see any major hurdles there. Especially since you've already worked through omf -> vtki conversion, this should mostly just be translating in the opposite direction.

From a few minutes of poking around the client library, it does indeed look similar to OMF - this should make my life a lot easier! 

Actually, I notice that there is an OMF importer... this makes me wonder if I should focus on creating a way to go from `vtki` to `omf` over in https://github.com/OpenGeoVis/omfvtk and put the LF View exporter/uploader over there? I'm thinking OMF might be a bit more stable than `lfview-resources-spatial` and that most people using `vtki`+ LF View would be `omfvtk` users to begin.

I suppose my remaining question is - should I focuse on building an exporter from `vtki` to `lfview-resources-spatial` or focus on buling the link from `vtki` back to `omf` so that users can both upload data to LF View and save projects locally.

**Comment by fwkoch** [comment]
Thanks @banesullivan - A few perspectives on this:

- For stability - you are probably correct that `omf` is more stable (with of course the major caveat that OMFv2 will be released by the end of the year with major updates: https://github.com/gmggroup/omf/wiki ).
- Regardless of stability, `vtki -> omf -> lfview` allows for more separation of concerns. The `lfview.client.convert.omf_to_view` function will be kept up-to-date with all changes both to `omf` and `lfview`, so by leveraging that, `vtki` only needs to worry about `omf`. This even means you don't even _have to_ build the exporter into `vtki` or `omfvtk` - you can just have an example in the docs that shows how to do this export by converting to OMF then using the `lfview` client.
- There are some advantages to using `lfview` directly, though:
  - An immediate benefit is that `lfview` has scene representation: https://lfview-resources-scene.readthedocs.io/en/latest/ - I'm not sure how useful this is with `vtki`, but possibly useful if you want to export exact camera angles, etc.
  - The flip side of "less stability" is "more flexibility, iteration, evolution, etc" - if there is a resource type OMF cannot represent, it is more likely to be added to the LFView API.
  - Direct connection to LFView could also be used to leverage data management. Rather than creating new OMF files every time you want to share, you can build up your project data through LFView and pull together the components from there as needed. (If that makes sense... happy to elaborate more...)

Not sure if those last three bullets are motivating enough. Lowest hanging fruit would definitely be extending `omfvtk` to export to OMF in addition to import. Then gaps in functionality could be identified and filled in with `lfview`.

**Comment by lperozzi** [comment]
Hi @banesullivan , as usual many thanks for your work.
Is there some news about the possibility to export `omf`or `vtki` to `LFview`?

**Comment by fwkoch** [comment]
Hey @lperozzi - I can't speak to support within `pyvista`/`omfvtk` (i.e. `vtki` -> `LFView` or `vtki` -> `omf` - Bane needs to follow up on those). However, the case of `omf` -> `LFView` is supported by the View Python client:

```
pip install lfview-api-client
```
Get an API key (https://lfview.com/generate_api_key - you must already be logged in), then - 

```python
from lfview.client import convert, Session
session = Session('your-api-key')
view = convert.omf_to_view('your_file.omf')
session.upload(view)
```

Maybe that helps a little, at least for now? Let me know if you have any questions or issues.

**Comment by lperozzi** [comment]
Hi @fwkoch. Many thanks. I will then try to import a `omf` project into as you describe

**Comment by banesullivan** [comment]
Hi @lperozzi, there's no news on exporting `pyvista` to `omf` at this time... however it's definitely up there on my list of todo's for [`omfvista`](https://github.com/OpenGeoVis/omfvista).

Considering how convienant the LFView API client is, this shouldn't be too dificult to implement when I have the time

**Comment by banesullivan** [comment]
Actually, there is a bit of a challenge: the conversion from OMF to VTK/PyVista is pretty clear, but the link from PyVista back to OMF isn't so clear, and there are several issues when trying to find appropriate data classes in OMF for all the types of data in VTK/PyVista because an OMF mesh type can often be converted to several different PyVista types

Unsupported features include:

- general `pyvista.StructuredGrid`s (curvilinear meshes) aren't supported in OMF v1
- general `pyvista.UnstructuredGrid`s aren't supported in OMF - they're a bit too complicated for OMF's current data types and a ton of checking would have to be implemented as `pyvista.UnstructuredGrid`s can have mixed cell types which isn't supported in OMF v1
- Tree meshes aren't supported in OMF v1 and hard to recover from PyVista as we use `UnstructuredGrid`s
- Texture mappings would need to be updated in PyVista to keep track of the origin, U, and V points in the `pyvista.Texture` class but that won't always work if the texture mapping doesn't correspond to a flat plane. Also, OMF might need to be updated to support NumPy arrays of textures (it might handle this now... I can't remember, but it was an issue at one point)

Conversions that would work right out of the box:

- `pyvista.PolyData` to either `SurfaceElement`, `LineSetElement`, or `PointSetElement` based on checking the cell types - if the mesh has mixed cell types it wouldn't be supported (for example `pyvista.PolyData` can have triangle, line, and vertice cells all in the same mesh)
- 3D `pyvista.UniformGrid`/`pyvista.RectilinearGrid` to OMF `VolumeElement`
- 2D `pyvista.UniformGrid`/`pyvista.RectilinearGrid` to OMF `SurfaceElement`

What I imagine for any sort of PyVista to LFView exporter would be that we take rendered scenes and extract the surface geometries of those datasets to have just triangulated surfaces, linesets, and PointSets which might mitigate these concerns.

Unfortunately, there's too much overlap between PyVista types and which OMF types they could be to quickly make a generalized conversion scheme - it would require tons of checking throughout cells and a lot of PyVista features simply wouldn't be supported in OMF and it'd be difficult to keep track of meta info like coordinate system references and texture mappings.


So in summary, we'd have to wait until OMF v2 and then see what makes sense... in all likliehood, `omfvista` will have to be completely overhauled for OMF v2 anyways so it might all work out nicely down the road.

**Comment by lperozzi** [comment]
@banesullivan thanks for the update

**Comment by banesullivan** [comment]
Closing. This won't be supported in PyVista. If there is a need/demand for it, it can be implemented in omfvista
